stan/math/rev/functor/coupled_ode_system.hpp

```diff
@@ -438,53 +112,62 @@ struct coupled_ode_system<F, var, var> {
    *    expected number of derivatives, N.
    */
   void operator()(const std::vector<double>& z, std::vector<double>& dz_dt,
-                  double t) const {
+                  double t) {
     using std::vector;
 
+    dz_dt.resize(size());
+
     // Run nested autodiff in this scope
     nested_rev_autodiff nested;
 
-    vector<var> y_vars(z.begin(), z.begin() + N_);
+    Eigen::Matrix<var, Eigen::Dynamic, 1> y_vars(N_);
+    for (size_t n = 0; n < N_; ++n)
+      y_vars.coeffRef(n) = z[n];
 
-    vector<var> dy_dt_vars = f_(t, y_vars, theta_nochain_, x_, x_int_, msgs_);
+    Eigen::Matrix<var, Eigen::Dynamic, 1> f_y_t_vars
+        = apply([&](auto&&... args) { return f_(t, y_vars, msgs_, args...); },
+                local_args_tuple_);
 
-    check_size_match("coupled_ode_system", "dz_dt", dy_dt_vars.size(), "states",
+    check_size_match("coupled_ode_system", "dy_dt", f_y_t_vars.size(), "states",
                      N_);
 
-    for (size_t i = 0; i < N_; i++) {
-      dz_dt[i] = dy_dt_vars[i].val();
-      dy_dt_vars[i].grad();
+    for (size_t i = 0; i < N_; ++i) {
+      dz_dt[i] = f_y_t_vars.coeffRef(i).val();
+      f_y_t_vars.coeffRef(i).grad();
 
-      for (size_t j = 0; j < N_; j++) {
-        // orders derivatives by equation (i.e. if there are 2 eqns
-        // (y1, y2) and 2 parameters (a, b), dy_dt will be ordered as:
-        // dy1_dt, dy2_dt, dy1_da, dy2_da, dy1_db, dy2_db
-        double temp_deriv = 0;
-        const size_t offset = N_ + N_ * j;
-        for (size_t k = 0; k < N_; k++) {
-          temp_deriv += z[offset + k] * y_vars[k].adj();
-        }
+      y_adjoints_ = y_vars.adj();
 
-        dz_dt[offset + i] = temp_deriv;
+      memset(args_adjoints_.data(), 0, sizeof(double) * num_args_vars);
+      apply(
+          [&](auto&&... args) {
+            accumulate_adjoints(args_adjoints_.data(), args...);
+          },
+          local_args_tuple_);
+
+      apply([&](auto&&... args) { zero_adjoints(args...); }, local_args_tuple_);
+
+      // No need to zero adjoints after last sweep
+      if (i + 1 < N_) {
+        nested.set_zero_all_adjoints();
```

Comment:
Not following why these adjoints need set to zero in different ways

Reply:
`nested` is a nested autodiff stack.

The things in `local_args_tuple` are actually on the main stack, so we can't use `set_zero_all_adjoints()` (this zeros more than just the adjoints we need) or `set_zero_nested_adjoints()` (this zeros things only on the nested stack).

It probably seems weird that some of the varis are on the main stack, but by doing this we only have to allocate them on the construction of `coupled_ode_system`, not every time `operator()` gets called. It's measurably faster for small models.

Reply:
I'm having a hard time following the logic of this code but if @wds15 or whomever says it's fine then I'm fine with it

Reply:
This code is engineered for maximal performance as it really pays off to that here.